src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp

```diff
@@ -82,315 +99,510 @@ enum FourHorsemen
+        uint32 GetData(uint32 type) const override
+        {
+            switch (type)
+            {
+                case DATA_MOVEMENT_FINISHED:
+                    return _myMovementFinished ? 1u : 0u;
+                case DATA_DEATH_TIME:
+                    return _timeDied;
+                case DATA_HORSEMEN_CHECK_ACHIEVEMENT_CREDIT:
+                {
+                    uint32 minTime = 0u, maxTime = 0u;
+                    for (Horseman boss : horsemen)
+                        if (Creature* cBoss = getHorsemanHandle(boss))
+                        {
+                            uint32 deathTime = cBoss->AI()->GetData(DATA_DEATH_TIME);
+                            if (!deathTime)
+                            {
+                                TC_LOG_WARN("scripts", "FourHorsemenAI: Checking for achievement credit but horseman %s is reporting not dead", cBoss->GetName().c_str());
+                                return 0u;
+                            }
+                            if (!minTime || deathTime < minTime)
+                                minTime = deathTime;
+                            if (!maxTime || deathTime > maxTime)
+                                maxTime = deathTime;
+                        }
+                        else
+                        {
+                            TC_LOG_WARN("scripts", "FourHorsemenAI: Checking for achievement credit but horseman with id %d is not present", uint32(boss));
+                            return 0u;
+                        }
+                    return (getMSTimeDiff(minTime, maxTime) <= 15 * IN_MILLISECONDS) ? 1u : 0u;
+                }
+                default:
+                    return 0u;
+            }
+        }
 
-        void Reset() override
+        void DoAction(int32 action) override
         {
-            if (!encounterActionReset)
-                DoEncounteraction(NULL, false, true, false);
+            switch (action)
+            {
+                case ACTION_BEGIN_MOVEMENT:
+                    me->GetMotionMaster()->MovePoint(1, _initialPath[0], true);
+                    break;
+                case ACTION_BEGIN_FIGHTING:
+                    me->SetCombatPulseDelay(5);
+                    BeginFighting();
+                    _ourMovementFinished = true;
+                    break;
+            }
+        }
 
-            instance->SetData(DATA_HORSEMEN0 + id, NOT_STARTED);
+        void CheckIsMovementFinished()
+        {
+            for (Horseman boss : horsemen)
+                if (Creature* cBoss = getHorsemanHandle(boss))
+                {
+                    if (cBoss->IsAlive() && !cBoss->AI()->GetData(DATA_MOVEMENT_FINISHED))
+                        return;
+                }
+                else
+                {
+                    TC_LOG_WARN("scripts", "FourHorsemenAI: Checking if movement is finished but horseman with id %d is not present", uint32(boss));
+                    return ResetEncounter();
```

Comment:
please use "return;" when the return type is void otherwise it might look like the called function is returning some kind of value.